src/CargoBuild/CargoBuildTask.cs

```diff
@@ -0,0 +1,437 @@
+                            {
+                                info.EnvironmentVariables.Add(envar.Key, envar.Value);
+                            }
+                        }
+                    }
+
+                    var process = Process.Start(info);
+                    int maxWait = 1000 * 60 * 15;
+
+                    var nameAndExtension = fileName.Substring(fileName.LastIndexOf('\\') + 1);
+                    Log.LogMessage(MessageImportance.Normal, $"\t\t\t\t\t\t*********** Start {nameAndExtension} logs ***********\n\n");
+
+                    using StreamReader errReader = process!.StandardError;
+                    _ = Log.LogMessagesFromStream(errReader, MessageImportance.Normal);
+
+                    using StreamReader outReader = process!.StandardOutput;
+                    _ = Log.LogMessagesFromStream(outReader, MessageImportance.Normal);
+                    Log.LogMessage(MessageImportance.Normal, $"\t\t\t\t\t\t*********** End {nameAndExtension} logs ***********\n\n");
+                    bool exited = process.WaitForExit(maxWait);
+                    if (!exited)
+                    {
+                        process.Kill();
+                        Log.LogError($"Killed process after max timeout reached : '{info.FileName}'");
+                        return -1;
+                    }
+
+                    return process.ExitCode;
+                });
+
+                processTask.Start();
+                return (ExitCode)await processTask;
+            }
+            catch (Exception ex)
+            {
+                Log.LogWarningFromException(ex);
+                return await Tasks.Task.FromResult(ExitCode.Failed);
+            }
+        }
+
+        private async Task<bool> DownloadRustupAsync()
+        {
+            if (File.Exists(_rustUpInitBinary) && await VerifyInitHashAsync())
+            {
+                return true;
+            }
+            else if (File.Exists(_rustUpInitBinary))
+            {
+                // If the hash doesn't match, that likely means there is a new version of rustup-init.exe available.
+                File.Delete(_rustUpInitBinary);
+            }
+
+            string rustupDownloadLink = _rustupDownloadLink;
+            Log.LogMessage(MessageImportance.Normal, $"Downloading -- {rustupDownloadLink}");
+            using (var client = new HttpClient())
+            {
+                Task<HttpResponseMessage> response = client.GetAsync(rustupDownloadLink);
+                if (!Directory.Exists(_rustInstallPath))
+                {
+                    Directory.CreateDirectory(_rustInstallPath);
+                }
+
+                using var responseStream = new FileStream(_rustUpInitBinary, FileMode.CreateNew);
+                HttpResponseMessage res = await response;
+                await res.Content.CopyToAsync(responseStream);
+            }
+
+            return await VerifyInitHashAsync();
```

Comment:
There's a race here since you're not downloading the files at the same time.

Should we retry this specific operation on failure?